src/config/include/config/Node.hpp

```diff
@@ -0,0 +1,444 @@
+     *
+     * ```yaml
+     * # Calling getPlural<int>("Bad","Bads") will fail
+     * # because Bads isn't a sequence.
+     * AnotherBadExample:
+     *   Bads: 3
+     * ```
+     *
+     * @tparam T the returned type. Must align with the return-type of F.
+     * @tparam F type of the callback/mapping function that maps from `Node` to `T`.
+     * @param singular the singular version of the key e.g. "Number"
+     * @param plural the plural version of the key e.g. "Numbers"
+     * @param f callback function mapping from the found node to a T instance. If not specified,
+     * will use `to<T>()`
+     * @return
+     */
+    template <typename T, typename F = std::function<T(const Node&)>>
+    std::vector<T> getPlural(
+        const std::string& singular, const std::string& plural, F&& f = [](const Node& n) {
+            return n.to<T>();
+        }) const;
+
+    friend std::ostream& operator<<(std::ostream& out, const Node& node) {
+        return out << YAML::Dump(node._yaml);
+    }
+
+private:
+    Node(const YAML::Node yaml, const Node* const parent, bool valid, std::string key)
+        : _yaml{yaml}, _parent{parent}, _valid{valid}, _key{std::move(key)} {}
+
+    Node(const YAML::Node yaml, const Node* const parent, std::string key)
+        : Node{yaml, parent, yaml, std::move(key)} {}
+
+
+    template <typename O, typename... Args>
+    static constexpr bool isNodeConstructible() {
+        // TODO: test of constness
+        return !std::is_trivially_constructible_v<O> &&
+            (std::is_constructible_v<O, Node&, Args...> ||
+             std::is_constructible_v<O, const Node&, Args...>);
+    }
+
+    template <typename O,
+              typename... Args,
+              typename = std::enable_if_t<isNodeConstructible<O, Args...>()>>
+    std::optional<O> _maybeImpl(Args&&... args) const {
+        return std::make_optional<O>(*this, std::forward<Args>(args)...);
+    }
+
+    template <typename O,
+              typename... Args,
+              typename = std::enable_if_t<
+                  !isNodeConstructible<O, Args...>() &&
+                  std::is_same_v<O, typename NodeConvert<O>::type> &&
+                  std::is_same_v<O,
+                                 decltype(NodeConvert<O>::convert(std::declval<Node>(),
+                                                                  std::declval<Args>()...))>>,
+              typename = void>
+    std::optional<O> _maybeImpl(Args&&... args) const {
+        return std::make_optional<O>(NodeConvert<O>::convert(*this, std::forward<Args>(args)...));
+    }
+
+    template <
+        typename O,
+        typename... Args,
+        typename = std::enable_if_t<
+            !isNodeConstructible<O, Args...>() &&
+            std::is_same_v<decltype(YAML::convert<O>::encode(std::declval<O>())), YAML::Node>>,
+        typename = void,
+        typename = void>
+    std::optional<O> _maybeImpl(Args&&... args) const {
```

Comment:
[question] How do you decide if a method goes in the header or the cpp file?

Reply:
so as a general rule anything templated pretty much has to go in the header....I moved whatever else I could and what was more than 1-2 lines of boilerplate.